File: src/main/resources/public/ts/services/SectionService.ts
import {idiom, ng, notify} from 'entcore';
import http from 'axios';
import {DataUtils} from "../utils/data";
import {Section} from "../models/Section";

export interface SectionService {
    list(formId: number) : Promise<any>;
    get(choiceId: number) : Promise<any>;
    save(section: Section) : Promise<any>;
    create(section: Section) : Promise<any>;
    update(section: Section) : Promise<any>;
    delete(sectionId: number) : Promise<any>;
}

export const sectionService: SectionService = {

    async list(formId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/forms/${formId}/sections`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.sectionService.list'));
            throw err;
        }
    },

    async get(sectionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/sections/${sectionId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.sectionService.get'));
            throw err;
        }
    },

    async save(section: Section) : Promise<any> {
        return section.id ? await this.update(section) : await this.create(section);
    },

    async create(section: Section) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/forms/${section.form_id}/sections`, section));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.sectionService.create'));
            throw err;
        }
    },

    async update(section: Section) : Promise<any> {
        try {
            return DataUtils.getData(await http.put(`/formulaire/forms/${section.form_id}/sections`, section));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.sectionService.update'));
            throw err;
        }
    },

    async delete(sectionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/sections/${sectionId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.sectionService.delete'));
            throw err;
        }
    }
};

export const SectionService = ng.service('SectionService',(): SectionService => sectionService);

File: formulaire/src/main/java/fr/openent/formulaire/service/SectionService.java
package fr.openent.formulaire.service;

import fr.openent.form.core.models.Section;
import fr.wseduc.webutils.Either;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public interface SectionService {
    /**
     * List all the sections of a specific form
     * @param formId form identifier
     * @param handler function handler returning JsonArray data
     */
    void list(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all the sections of a specific form
     * @param formId form identifier
     */
    Future<JsonArray> list(String formId);

    /**
     * Get a specific section by id
     * @param sectionId section identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String sectionId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a section in a specific form
     * @param section Section data
     * @param formId form identifier
     */
    Future<JsonObject> create(Section section, String formId);

    /**
     * Update specific sections
     * @param formId form identifier
     * @param sections JsonArray data
     */
    Future<JsonArray> update(String formId, JsonArray sections);

    /**
     * Delete a specific section
     * @param section section
     * @param handler function handler returning JsonObject data
     */
    void delete(JsonObject section, Handler<Either<String, JsonObject>> handler);

    /**
     * Check validity of a specific section
     * @param section Section data
     */
    Future<Boolean> isTargetValid(Section section);
}
